src/main/scala/firrtl/transforms/TopWiring.scala

```diff
@@ -0,0 +1,247 @@
+    * 2. Connect these ports to ports of instances that are parents to some number of target wires
+    */
+  // TODO Make code more clear
+  private def onModule(prefix: String, sources: Map[String, Seq[(ComponentName, Type, Boolean, InstPath)]])
+                      (module: DefModule): DefModule = {
+    sources.get(module.name) match {
+      case Some(p) => 
+        val newPorts = p.map{ case (ComponentName(cname,_), tpe, _ , path) => Port(NoInfo, prefix + path.mkString("_"), Output, tpe)}
+        
+
+        // Add connections to Module
+        module match {
+          case m: Module =>
+            val connections: Seq[Connect] = p.map { case (ComponentName(cname,_), _, _ , path) =>
+                val modRef = WRef(prefix + path.mkString("_"))
+                path.size match {
+                   case 1 => {
+                       val leafRef = WRef(path.head.mkString("_"))
+                       Connect(NoInfo, modRef, leafRef)
+                   }
+                   case _ =>  {
+                       val instRef = WSubField(WRef(path.head), prefix + path.tail.mkString("_"))
+                       Connect(NoInfo, modRef, instRef)
+                  }
+                }
+            } 
+            m.copy(ports = m.ports ++ newPorts, body = Block(Seq(m.body) ++ connections ))
+          case e: ExtModule =>
+            e.copy(ports = e.ports ++ newPorts)
+      }
+      case None => module // unchanged if no paths
+    }
+  }
+
+  private def fixupCircuit(circuit: Circuit): Circuit = {
+    val passes = Seq(
+      InferTypes,
+      ResolveKinds,
+      ResolveGenders
+    )
+    passes.foldLeft(circuit) { case (c: Circuit, p: Pass) => p.run(c) }
+  }
+
+
+  def TopWiringDummyOutputFilesFunction(prefix: String, dir: String, mapping: Seq[((ComponentName, Type, Boolean, InstPath), Int)], state: CircuitState): CircuitState = {
+     state  
+  }
+
+  def TopWiringTestOutputFilesFunction(prefix: String, dir: String, mapping: Seq[((ComponentName, Type, Boolean, InstPath), Int)], state: CircuitState): CircuitState = {
+     val testOutputFile = new PrintWriter(new File(dir, "TopWiringOutputTest.txt" )) 
+     mapping map {
+          case ((_, tpe, _, path), index) => {
+            val portwidth = tpe match { case GroundType(IntWidth(w)) => w }
+            val portnum = index
+            val portname = prefix + path.mkString("_")
+            testOutputFile.append(s"new top level port $portnum : $portname, with width $portwidth \n")
+          }
+     }
+     testOutputFile.close()
+     state 
+  }
+
+
+  def execute(state: CircuitState): CircuitState = {
+
+    lazy val outputTuple: Option[(String, (String,String,Seq[((ComponentName, Type, Boolean, InstPath), Int)], CircuitState) => CircuitState)] = state.annotations.collectFirst { case TopWiringOutputFilesAnnotation(td,of) => (td, of) }
+    lazy val prfix: Option[String] = state.annotations.collectFirst { case TopWiringPrefixAnnotation(pr) => pr }
```

Comment:
Same issue with `collectFirst` here. How are you going to deal with multiple prefixes for different sources that you want to wire out?